invenio_oaiharvester/api.py

```diff
@@ -0,0 +1,107 @@
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+from sickle import Sickle
+from invenio_oaiharvester.errors import (NameOrUrlMissing, WrongDateCombination,
+                                         NAME_OR_URL_ERROR_MSG, WRONG_DATES_ERROR_MSG)
+from invenio_oaiharvester.models import OaiHARVEST
+
+
+def list_records(metadata_prefix, from_date=None, until_date=None, url=None, name=None, setSpec=None):
+    """
+    Harvest records from an OAI repo, based on datestamp and/or set parameters.
+    :param metadata_prefix: The prefix for the metadata return (e.g. 'arXiv') (required).
+    :param from_date: The lower bound date for the harvesting (optional).
+    :param until_date: The upper bound date for the harvesting (optional).
+    :param url: The The url to be used to create the endpoint.
+    :param name: The name of the OaiHARVEST object that we want to use to create the endpoint.
+    :param setSpec: The 'set' criteria for the harvesting (optional).
+    :return: A list of harvested records.
+    """
+    request = make_endpoint_request(url, name)
+
+    if from_date is None:   # then we are using object.lastrun
+        obj = get_oaiharvest_obj(name)
+        from_date = obj.lastrun
+        update_lastrun(obj)
+
+    dates = {'from': from_date}\
+        if until_date is None \
+        else {'from': from_date, 'until': until_date}
+
+    # Sanity check
+    if from_date > until_date:
+        raise WrongDateCombination(WRONG_DATES_ERROR_MSG)
+
+    return list(request.ListRecords(metadataPrefix=metadata_prefix, set=setSpec, **dates))
+
+
+def get_records(metadata_prefix, identifiers, url=None, name=None):
+    """
+    Harvest specific records from an OAI repo, based on their unique identifiers.
+    :param metadata_prefix: The prefix for the metadata return (e.g. 'arXiv') (required).
+    :param identifiers: A list of unique identifiers for records to be harvested.
+    :param url: The The url to be used to create the endpoint.
+    :param name: The name of the OaiHARVEST object that we want to use to create the endpoint.
+    :return: A list of harvested records.
+    """
+    request = make_endpoint_request(url, name)
+
+    records = []
+    for identifier in identifiers:
+        arguments = {'identifier': identifier, 'metadataPrefix': metadata_prefix}
+        records.append(request.GetRecord(**arguments))
+
+    return records
+
+
+def make_endpoint_request(url=None, name=None):
+    """
+    Uses a provided url or name to create an endpoint for the harvest requests
+    (OAI interface). The function needs either the url or the name.
+    :param url: The url to be used to create the endpoint.
+    :param name: The name of the OaiHARVEST object that we want to use to create the endpoint.
+    :return: An endpoint object for future http requests.
+    """
+    if url:
+        return Sickle(url)
+    elif name:
+        obj = get_oaiharvest_obj(name)
```

Comment:
To avoid potentially calling `get_oaiharvest_obj` twice, e.g. 2x database queries, (here and in `list_records`), it could be better if this function simply just accepts `url` and we do the `get_oaiharvest_obj` outside of this function.